rose/server/score.py

```diff
@@ -27,49 +27,65 @@ def process(players, track):
             if player.x < config.matrix_width - 1:
                 player.x += 1
 
-    # Now handle obstacles, preferring players in their own lane.
+    # Proccess the players by order, first the ones in lane and then
+    # the ones out of lane, this ensure the car in lane will have
+    # priority when picking pinguins and in case of collisions.
 
     sorted_players = sorted(six.itervalues(players),
                             key=lambda p: 0 if p.in_lane() else 1)
     positions = set()
 
+    # Now handle obstacles, preferring players in their own lane.
+
     for player in sorted_players:
-        player.score += config.score_move_forward
         obstacle = track.get(player.x, player.y)
-        if obstacle == obstacles.CRACK:
-            if player.action != actions.JUMP:
-                track.clear(player.x, player.y)
-                player.y += 1
-                player.score += config.score_move_backward * 2
-            else:
-                player.score += config.score_jump
+
+        if obstacle == obstacles.NONE:
+            # Move forlward
+            player.score += config.score_move_forward
+
         elif obstacle in (obstacles.TRASH,
                           obstacles.BIKE,
                           obstacles.BARRIER):
-            if player.action not in (actions.LEFT, actions.RIGHT):
+            # Move back
+            track.clear(player.x, player.y)
+            player.y += 1
+            player.score += config.score_move_backward
+
+        elif obstacle == obstacles.CRACK:
+            if player.action == actions.JUMP:
+                # Move forlward
```

Comment:
Maybe:

    # Move forward leaving the obstacle on the track

Reply:
done